During certbot-auto upgrade: update venv if it exists

gboudreau opened the pull request:
When `certbot-auto` detects it should upgrade, it used to just delete the whole venv folder, and re-create it.
This caused issues with manually installed plugins that just disappeared.

With this change, instead of deleting the whole venv folder, we just delete a few files/folders from it, to force them to get re-installed, and keep the rest as-is.

Fixes #4037

gboudreau commented:
Request for comments: is "updating" the virtualenv in this manner "safe"?
When/how could it cause issues..?

erikrose commented on letsencrypt-auto-source/letsencrypt-auto:
```diff
@@ -613,9 +613,14 @@ if [ "$1" = "--le-auto-phase2" ]; then
+    if [ -d "$VENV_PATH" ]; then
+        echo "Updating virtual environment..."
+        # The following will insure the latest Python core files and standard libraries (setuptools, pip) gets installed, while custom plugins and libraries installed in site-packages/ are kept as-is
+        rm -rf "$VENV_PATH"/bin/pip "$VENV_PATH"/bin/pip2 "$VENV_PATH"/bin/pip2.7 "$VENV_PATH"/bin/python "$VENV_PATH"/bin/python2 "$VENV_PATH"/bin/python2.7 "$VENV_PATH"/include/python2.7 "$VENV_PATH"/lib/python2.7/distutils "$VENV_PATH"/lib/python2.7/*-packages/easy_install.* "$VENV_PATH"/lib/python2.7/*-packages/pip "$VENV_PATH"/lib/python2.7/*-packages/pip-*.dist-info
```
I think this approach will bite us in the future if the logic in virtualenv changes. Did the virtualenv authors make any promise that future versions will behave the same way in response to these deletions?

Also, we won't get the latest setuptools or pip in any case, as those are pinned in pipstrap. The safest path, it seems to me, is to leave the venv alone on upgrade.

gboudreau replied:
OK; removed that (in 68071be56b4ab8a20416c61a15aae1370fb3682d below).

erikrose commented:
(Obviously, the test failures will need to be fixed before merging.)

gboudreau commented:
Hum. It only fails on trusty only... No idea why.
I rebased from master, and squashed my commits, to insure that wasn't the problem.

Any ideas?
https://travis-ci.org/certbot/certbot/jobs/193163763#L781

L297 in auto_test.py is:

                # Test when a phase-1 upgrade is not needed but a phase-2
                # upgrade is:
                set_le_script_version(venv_dir, '0.0.1')
                out, err = run_letsencrypt_auto()

That seems to indicate there is a problem with keeping the venv after all... :|

bmw commented:
Output from `letsencrypt-auto` here seems to be:
```
Updating virtual environment...
Installing Python packages...

Traceback (most recent call last):
  File "/tmp/le-test-gp1rbp/letsencrypt/bin/pip", line 9, in <module>
    load_entry_point('pip==8.0.3', 'console_scripts', 'pip')()
  File "/tmp/le-test-gp1rbp/letsencrypt/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 549, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/tmp/le-test-gp1rbp/letsencrypt/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2542, in load_entry_point
    return ep.load()
  File "/tmp/le-test-gp1rbp/letsencrypt/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2202, in load
    return self.resolve()
  File "/tmp/le-test-gp1rbp/letsencrypt/local/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2208, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/__init__.py", line 11, in <module>
    from pip.vcs import git, mercurial, subversion, bazaar  # noqa
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/vcs/mercurial.py", line 9, in <module>
    from pip.download import path_to_url
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/download.py", line 22, in <module>
    from pip._vendor import requests, six
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/__init__.py", line 58, in <module>
    from . import utils
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/utils.py", line 25, in <module>
    from .compat import parse_http_list as _parse_list_header
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/compat.py", line 7, in <module>
    from .packages import chardet
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/packages/__init__.py", line 3, in <module>
    from . import urllib3
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/__init__.py", line 16, in <module>
    from .connectionpool import (
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/connectionpool.py", line 33, in <module>
    from .connection import (
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/connection.py", line 41, in <module>
    from .util import (
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/__init__.py", line 5, in <module>
    from .response import is_fp_closed
  File "/tmp/le-test-gp1rbp/letsencrypt/lib/python2.7/site-packages/pip/_vendor/requests/packages/urllib3/util/response.py", line 4, in <module>
    from ..exceptions import HeaderParsingError
ImportError: cannot import name HeaderParsingError
Traceback (most recent call last):
  File "/tmp/tmp.GodkfXfrPG/pipstrap.py", line 146, in <module>
    exit(main())
  File "/tmp/tmp.GodkfXfrPG/pipstrap.py", line 133, in main
    shell=True)
  File "/usr/lib/python2.7/subprocess.py", line 573, in check_output
    raise CalledProcessError(retcode, cmd, output=output)
subprocess.CalledProcessError: Command 'pip install --no-index --no-deps -U /tmp/pipstrap-2jo9uu/pip-8.0.3.tar.gz /tmp/pipstrap-2jo9uu/setuptools-20.2.2.tar.gz /tmp/pipstrap-2jo9uu/wheel-0.29.0.tar.gz' returned non-zero exit status 1
```

erikrose commented:
Have you tried running plain old master on trusty (i.e. submitting a basically empty PR)? Maybe something changed remotely.

bmw commented:
Other PRs that don't affect `letsencrypt-auto` are still passing. I'll save the log and try kicking Travis to see if it's a Heisenbug.

erikrose commented:
Here's an idea: certbot (and letsencrypt-auto/pieces/letsencrypt-auto-requirements.txt) install requests. Ordinarily, pip is supposed to use the copy of requests in its vendor lib, which it certainly does when operating from an empty venv. However, then when operating from a prepopulated venv, maybe there's something weird about trusty (the order of `sys.path`?) that makes it find the non-pip version of requests.

ohemorange commented:
Hey @zjs, this is probably necessary for getting other plugins working. There's a bunch of work that's already been done on it, but there are issues that could use some ironing out. Think you'd be able to take this on?

bmw commented:
This PR hasn't seen any progress in a while and this approach seems to cause issues on some systems as seen by test failures. We certainly want to see a solution to #1279 though so if you continue working on it, please open a new PR and we'll happily review.
